File: jscomp/others/js_json_decode.ml
(* Copyright (C) 2015-2016 Bloomberg Finance L.P.
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * In addition to the permissions granted to you by the LGPL, you may combine
 * or link a "work that uses the Library" with a publicly distributed version
 * of this file to produce a combined library or application, then distribute
 * that combined work under the terms of your choosing, with no requirement
 * to comply with the obligations normally placed on you by section 4 of the
 * LGPL version 3 (or the corresponding section of a later version of the LGPL
 * should you choose to use a later version).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)


open Bs.Result

type 'a decoder = Js.json -> ('a, string) result

let boolean json = 
  if Js.typeof json = "boolean"
  then Ok (Obj.magic (json : Js.json) : Js.boolean)
  else Error ("Expected boolean, got " ^ Js_json.stringify json)

let float json = 
  if Js.typeof json = "number" 
  then Ok (Obj.magic (json : Js.json) : float)
  else Error ("Expected number, got " ^ Js_json.stringify json)

let int json = 
  match float json with
  | Ok float ->
    if Js.Float.isInteger float
    then Ok (Obj.magic (float : float) : int)
    else Error ("Expected integer, got " ^ Js_json.stringify json)
  | Error message -> Error message

let string json = 
  if Js.typeof json = "string" 
  then Ok (Obj.magic (json:Js.json) : string)
  else Error ("Expected string, got " ^ Js_json.stringify json)

let nullable decode json =
  if (Obj.magic json : 'a Js.null) == Js.null
  then Ok Js.null
  else
    match decode json with
    | Ok value -> Ok (Js.Null.return value)
    | Error message -> Error message

(* TODO: remove this? *)
let nullAs value json = 
  if (Obj.magic json : 'a Js.null) == Js.null
  then Ok value
  else Error ("Expected null, got " ^ Js_json.stringify json)

let array decode json = 
  if Js_array.isArray json
  then begin
    (* TODO: Seriously rethink this for better balance between readability and perf *)
    let source = (Obj.magic (json : Js.json) : Js.json array) in
    let l = Js.Array.length source in

    if l = 0 then Ok [||]

    else begin
      let target = Array.make l (Obj.magic 0) in
      let break = ref false in
      let result = ref (Ok target) in
      let i = ref 0 in
      while not !break do
        if !i >= l then break := true
        else
          match decode (Array.unsafe_get source !i) with
          | Ok value ->
            Array.set target !i value;
            i := !i + 1
          | Error message ->
            result := Error message;
            break := true
      done;
      !result
    end
  end
  else Error ("Expected array, got " ^ Js_json.stringify json)

let dict decode json = 
  if Js.typeof json = "object" && 
      not (Js_array.isArray json) && 
      not ((Obj.magic json : 'a Js.null) == Js.null)
  then begin
    (* TODO: Seriously rethink this for better balance between readability and perf *)
    let source = (Obj.magic (json : Js.json) : Js.json Js_dict.t) in
    let keys = Js_dict.keys source in
    let l = Js.Array.length keys in

    if l = 0 then Ok (Js_dict.empty ())

    else begin
      let target = Js_dict.empty () in
      let break = ref false in
      let result = ref (Ok target) in
      let i = ref 0 in
      while not !break do
        if !i >= l then break := true
        else
          let key = (Array.unsafe_get keys !i) in
          match decode (Js_dict.unsafeGet source key) with
          | Ok value ->
            Js_dict.set target key value;
            i := !i + 1
          | Error message ->
            result := Error message;
            break := true
      done;
      !result
    end
  end
  else Error ("Expected object, got " ^ Js_json.stringify json)

let field key decode json =
  if Js.typeof json = "object" && 
      not (Js_array.isArray json) && 
      not ((Obj.magic json : 'a Js.null) == Js.null)
  then begin
    let dict = (Obj.magic (json : Js.json) : Js.json Js.Dict.t) in
    match Js.Dict.get dict key with
    | Some value -> decode value
    | None -> Error ("Expected field '" ^ key ^ "'")
  end
  else Error ("Expected object, got " ^ Js_json.stringify json)

let optional decode json =
  match decode json with
  | Ok value -> Ok (Some value)
  | Error message -> Ok None

File: lib/js/js_json_decode.js
'use strict';

var Block      = require("./block");
var Curry      = require("./curry");
var Caml_array = require("./caml_array");

function $$boolean(json) {
  if (typeof json === "boolean") {
    return /* Ok */Block.__(0, [json]);
  } else {
    return /* Error */Block.__(1, ["Expected boolean, got " + JSON.stringify(json)]);
  }
}

function $$float(json) {
  if (typeof json === "number") {
    return /* Ok */Block.__(0, [json]);
  } else {
    return /* Error */Block.__(1, ["Expected number, got " + JSON.stringify(json)]);
  }
}

function $$int(json) {
  var match = $$float(json);
  if (match.tag) {
    return /* Error */Block.__(1, [match[0]]);
  } else {
    var $$float$1 = match[0];
    if (Number.isInteger($$float$1)) {
      return /* Ok */Block.__(0, [$$float$1]);
    } else {
      return /* Error */Block.__(1, ["Expected integer, got " + JSON.stringify(json)]);
    }
  }
}

function string(json) {
  if (typeof json === "string") {
    return /* Ok */Block.__(0, [json]);
  } else {
    return /* Error */Block.__(1, ["Expected string, got " + JSON.stringify(json)]);
  }
}

function nullable(decode, json) {
  if (json === null) {
    return /* Ok */Block.__(0, [null]);
  } else {
    var match = Curry._1(decode, json);
    if (match.tag) {
      return /* Error */Block.__(1, [match[0]]);
    } else {
      return /* Ok */Block.__(0, [match[0]]);
    }
  }
}

function nullAs(value, json) {
  if (json === null) {
    return /* Ok */Block.__(0, [value]);
  } else {
    return /* Error */Block.__(1, ["Expected null, got " + JSON.stringify(json)]);
  }
}

function array(decode, json) {
  if (Array.isArray(json)) {
    var l = json.length;
    if (l) {
      var target = Caml_array.caml_make_vect(l, 0);
      var $$break = /* false */0;
      var result = /* Ok */Block.__(0, [target]);
      var i = 0;
      while(!$$break) {
        if (i >= l) {
          $$break = /* true */1;
        } else {
          var match = Curry._1(decode, json[i]);
          if (match.tag) {
            result = /* Error */Block.__(1, [match[0]]);
            $$break = /* true */1;
          } else {
            Caml_array.caml_array_set(target, i, match[0]);
            i = i + 1 | 0;
          }
        }
      };
      return result;
    } else {
      return /* Ok */Block.__(0, [/* array */[]]);
    }
  } else {
    return /* Error */Block.__(1, ["Expected array, got " + JSON.stringify(json)]);
  }
}

function dict(decode, json) {
  if (typeof json === "object" && !Array.isArray(json) && json !== null) {
    var keys = Object.keys(json);
    var l = keys.length;
    if (l) {
      var target = { };
      var $$break = /* false */0;
      var result = /* Ok */Block.__(0, [target]);
      var i = 0;
      while(!$$break) {
        if (i >= l) {
          $$break = /* true */1;
        } else {
          var key = keys[i];
          var match = Curry._1(decode, json[key]);
          if (match.tag) {
            result = /* Error */Block.__(1, [match[0]]);
            $$break = /* true */1;
          } else {
            target[key] = match[0];
            i = i + 1 | 0;
          }
        }
      };
      return result;
    } else {
      return /* Ok */Block.__(0, [{ }]);
    }
  } else {
    return /* Error */Block.__(1, ["Expected object, got " + JSON.stringify(json)]);
  }
}

function field(key, decode, json) {
  if (typeof json === "object" && !Array.isArray(json) && json !== null) {
    var match = json[key];
    if (match !== undefined) {
      return Curry._1(decode, match);
    } else {
      return /* Error */Block.__(1, ["Expected field '" + (key + "'")]);
    }
  } else {
    return /* Error */Block.__(1, ["Expected object, got " + JSON.stringify(json)]);
  }
}

function optional(decode, json) {
  var match = Curry._1(decode, json);
  if (match.tag) {
    return /* Ok */Block.__(0, [/* None */0]);
  } else {
    return /* Ok */Block.__(0, [/* Some */[match[0]]]);
  }
}

exports.$$boolean = $$boolean;
exports.$$float   = $$float;
exports.$$int     = $$int;
exports.string    = string;
exports.nullable  = nullable;
exports.nullAs    = nullAs;
exports.array     = array;
exports.dict      = dict;
exports.field     = field;
exports.optional  = optional;
/* No side effect */
